java-checks/src/main/java/org/sonar/java/checks/CallOuterPrivateMethodCheck.java

```diff
@@ -0,0 +1,103 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.HashMultiset;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Multiset;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.tag.Tag;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+
+@Rule(
+  key = "S3398",
+  name = "\"private\" methods called only by inner classes should be moved to those classes",
+  priority = Priority.MAJOR,
+  tags = {Tag.CONFUSING})
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.READABILITY)
+@SqaleConstantRemediation("5min")
+public class CallOuterPrivateMethodCheck extends SubscriptionBaseVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.CLASS, Tree.Kind.INTERFACE);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+    Symbol.TypeSymbol classSymbol = classTree.symbol();
+    if (isInnerClass(classSymbol)) {
+      MethodInvocationVisitor methodInvocationVisitor = new MethodInvocationVisitor(classSymbol);
+      classTree.accept(methodInvocationVisitor);
+      methodInvocationVisitor.checkUsages();
+    }
+  }
+
+  private static boolean isInnerClass(Symbol symbol) {
+    return symbol.owner().isTypeSymbol();
+  }
+
+  private class MethodInvocationVisitor extends BaseTreeVisitor {
+    private final Symbol.TypeSymbol classSymbol;
+    private final Multiset<Symbol> usages = HashMultiset.create();
+
+    public MethodInvocationVisitor(Symbol.TypeSymbol classSymbol) {
+      this.classSymbol = classSymbol;
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      Symbol symbol = tree.symbol();
+      if (isPrivateMethodOfOuterClass(symbol)) {
+        usages.add(symbol);
+      }
+      super.visitMethodInvocation(tree);
+    }
+
+    private boolean isPrivateMethodOfOuterClass(Symbol symbol) {
+      return symbol.isPrivate() && symbol.owner() == classSymbol.owner();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)